File: packages/tx402-python/src/tx402/fingerprint.py
"""Secret-free request fingerprinting (SEC-009)."""

from __future__ import annotations

import hashlib
import posixpath
from urllib.parse import SplitResult, quote, urlsplit, urlunsplit

from tx402.canonical_json import canonicalize_json

REQUEST_FINGERPRINT_DOMAIN = "tx402-request-fingerprint-v1\n"


def _sha256(value: bytes) -> str:
    return f"sha256:{hashlib.sha256(value).hexdigest()}"


def normalize_fingerprint_url(value: str) -> str:
    """Normalize a URL while dropping secret user-info and the non-transmitted fragment."""
    parsed = urlsplit(value)
    if not parsed.scheme or parsed.hostname is None:
        raise ValueError("Fingerprint URL must be absolute")
    scheme = parsed.scheme.lower()
    hostname = parsed.hostname.encode("idna").decode("ascii").lower()
    if ":" in hostname:
        hostname = f"[{hostname}]"
    port = parsed.port
    if port is not None and not (
        (scheme == "https" and port == 443) or (scheme == "http" and port == 80)
    ):
        hostname = f"{hostname}:{port}"
    path = parsed.path or "/"
    trailing = path.endswith("/")
    path = posixpath.normpath(path)
    if not path.startswith("/"):
        path = f"/{path}"
    if trailing and path != "/":
        path = f"{path}/"
    path = quote(path, safe="/%:@-._~!$&'()*+,;=")
    return urlunsplit(SplitResult(scheme, hostname, path, parsed.query, ""))


def digest_request_body(body: str | bytes | None) -> str:
    """Digest raw request-body bytes; ``None`` is the empty byte sequence."""
    raw = b"" if body is None else body.encode("utf-8") if isinstance(body, str) else body
    return _sha256(raw)


def fingerprint_request(
    *, method: str, url: str, body: str | bytes | None, challenge_hash: str
) -> str:
    """Bind method, normalized URL, body digest, and challenge hash."""
    document = {
        "bodyHash": digest_request_body(body),
        "challengeHash": challenge_hash,
        "method": method.upper(),
        "url": normalize_fingerprint_url(url),
    }
    payload = f"{REQUEST_FINGERPRINT_DOMAIN}{canonicalize_json(document)}".encode()
    return _sha256(payload)

File: packages/tx402/src/core/fingerprint.ts
/** Secret-free request fingerprinting (SEC-009). */

import { createHash } from "node:crypto";

import { canonicalizeJson } from "./canonical-json.js";

export const REQUEST_FINGERPRINT_DOMAIN = "tx402-request-fingerprint-v1\n";

export interface RequestFingerprintInput {
  readonly method: string;
  readonly url: string | URL;
  readonly body: string | Uint8Array | null;
  readonly challengeHash: string;
}

function sha256(bytes: string | Uint8Array): string {
  return `sha256:${createHash("sha256").update(bytes).digest("hex")}`;
}

/**
 * Normalize the HTTP URL component while removing secret user-info and the non-transmitted
 * fragment. WHATWG URL canonicalizes scheme/host case, default ports, dot segments, and IDNs.
 */
export function normalizeFingerprintUrl(input: string | URL): string {
  const url = new URL(input);
  url.username = "";
  url.password = "";
  url.hash = "";
  return url.href;
}

/** Digest raw request-body bytes. `null` and an empty body intentionally share a digest. */
export function digestRequestBody(body: string | Uint8Array | null): string {
  return sha256(body === null ? new Uint8Array() : body);
}

/**
 * Bind method, normalized secret-free URL, raw body digest, and challenge hash.
 * The domain-separated canonical JSON byte construction is frozen by shared vectors.
 */
export function fingerprintRequest(input: RequestFingerprintInput): string {
  const document = {
    bodyHash: digestRequestBody(input.body),
    challengeHash: input.challengeHash,
    method: input.method.toUpperCase(),
    url: normalizeFingerprintUrl(input.url),
  };
  return sha256(`${REQUEST_FINGERPRINT_DOMAIN}${canonicalizeJson(document)}`);
}
